File: cobble-binding/cobble-cpp/cpp/structured/priority_queue.cpp
#include <cobble/structured/db.hpp>
#include <cobble/structured/priority_queue.hpp>
#include <cobble/structured/single_db.hpp>

#include <utility>

#include "../detail/error.hpp"
#include "detail/convert.hpp"
#include "detail/impl.hpp"

namespace cobble::structured {
namespace {

template <typename T>
void Require(const std::unique_ptr<T> &value, const char *message) {
  if (!value)
    throw Error(ErrorCode::kInvalidState, message);
}

template <typename T>
void Require(const std::shared_ptr<T> &value, const char *message) {
  if (!value)
    throw Error(ErrorCode::kInvalidState, message);
}

template <typename T>
void RequireExclusiveOwner(const std::shared_ptr<T> &value) {
  if (value.use_count() != 1) {
    throw Error(ErrorCode::kInvalidState,
                "priority queue creation requires releasing every scan "
                "cursor, schema builder, and priority queue first");
  }
}

std::pair<bool, std::size_t>
NativeLimit(std::optional<std::size_t> limit) noexcept {
  return {limit.has_value(), limit.value_or(0)};
}

} // namespace

OwnedPriorityQueueEntry::OwnedPriorityQueueEntry(
    std::unique_ptr<Impl> impl) noexcept
    : impl_(std::move(impl)) {}
OwnedPriorityQueueEntry::OwnedPriorityQueueEntry(
    OwnedPriorityQueueEntry &&) noexcept = default;
OwnedPriorityQueueEntry &OwnedPriorityQueueEntry::operator=(
    OwnedPriorityQueueEntry &&) noexcept = default;
OwnedPriorityQueueEntry::~OwnedPriorityQueueEntry() = default;

BytesView OwnedPriorityQueueEntry::Key() const {
  Require(impl_, "OwnedPriorityQueueEntry has been moved from");
  return detail::ToView(detail::Translate([&] {
    return structured_ffi::native_priority_queue_batch_key(*impl_->native, 0);
  }));
}

BytesView OwnedPriorityQueueEntry::Value() const {
  Require(impl_, "OwnedPriorityQueueEntry has been moved from");
  return detail::ToView(detail::Translate([&] {
    return structured_ffi::native_priority_queue_batch_value(*impl_->native,
                                                             0);
  }));
}

OwnedPriorityQueueBatch::OwnedPriorityQueueBatch(
    std::unique_ptr<Impl> impl) noexcept
    : impl_(std::move(impl)) {}
OwnedPriorityQueueBatch::OwnedPriorityQueueBatch(
    OwnedPriorityQueueBatch &&) noexcept = default;
OwnedPriorityQueueBatch &OwnedPriorityQueueBatch::operator=(
    OwnedPriorityQueueBatch &&) noexcept = default;
OwnedPriorityQueueBatch::~OwnedPriorityQueueBatch() = default;

std::size_t OwnedPriorityQueueBatch::Size() const noexcept {
  return impl_ ? structured_ffi::native_priority_queue_batch_size(*impl_->native)
               : 0;
}

PriorityQueueEntryView
OwnedPriorityQueueBatch::Entry(std::size_t index) const {
  Require(impl_, "OwnedPriorityQueueBatch has been moved from");
  return PriorityQueueEntryView{
      detail::ToView(detail::Translate([&] {
        return structured_ffi::native_priority_queue_batch_key(*impl_->native,
                                                               index);
      })),
      detail::ToView(detail::Translate([&] {
        return structured_ffi::native_priority_queue_batch_value(
            *impl_->native, index);
      }))};
}

PriorityQueue::PriorityQueue(std::unique_ptr<Impl> impl) noexcept
    : impl_(std::move(impl)) {}
PriorityQueue::PriorityQueue(PriorityQueue &&) noexcept = default;
PriorityQueue &PriorityQueue::operator=(PriorityQueue &&) noexcept = default;
PriorityQueue::~PriorityQueue() = default;

std::string PriorityQueue::ColumnFamily() const {
  Require(impl_, "PriorityQueue has been moved from");
  return std::string(
      structured_ffi::native_structured_priority_queue_column_family(
          *impl_->native));
}

void PriorityQueue::Offer(BucketId bucket, BytesView key,
                          BytesView value) const {
  Require(impl_, "PriorityQueue has been moved from");
  detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_offer(
        *impl_->native, bucket, detail::RustBytes(key),
        detail::RustBytes(value));
  });
}

void PriorityQueue::Delete(BucketId bucket, BytesView key) const {
  Require(impl_, "PriorityQueue has been moved from");
  detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_delete(
        *impl_->native, bucket, detail::RustBytes(key));
  });
}

std::optional<OwnedPriorityQueueEntry>
PriorityQueue::Peek(BucketId bucket) const {
  Require(impl_, "PriorityQueue has been moved from");
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_peek(*impl_->native,
                                                                 bucket);
  });
  if (structured_ffi::native_priority_queue_batch_size(*native) == 0)
    return std::nullopt;
  OwnedPriorityQueueEntry entry(
      std::make_unique<OwnedPriorityQueueEntry::Impl>(std::move(native)));
  return std::optional<OwnedPriorityQueueEntry>(std::move(entry));
}

std::optional<OwnedPriorityQueueEntry>
PriorityQueue::Poll(BucketId bucket) const {
  Require(impl_, "PriorityQueue has been moved from");
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_poll(*impl_->native,
                                                                 bucket);
  });
  if (structured_ffi::native_priority_queue_batch_size(*native) == 0)
    return std::nullopt;
  OwnedPriorityQueueEntry entry(
      std::make_unique<OwnedPriorityQueueEntry::Impl>(std::move(native)));
  return std::optional<OwnedPriorityQueueEntry>(std::move(entry));
}

OwnedPriorityQueueBatch
PriorityQueue::PeekBatch(BucketId bucket,
                         std::optional<std::size_t> limit) const {
  Require(impl_, "PriorityQueue has been moved from");
  const auto [has_limit, native_limit] = NativeLimit(limit);
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_peek_batch(
        *impl_->native, bucket, has_limit, native_limit);
  });
  return OwnedPriorityQueueBatch(
      std::make_unique<OwnedPriorityQueueBatch::Impl>(std::move(native)));
}

OwnedPriorityQueueBatch
PriorityQueue::PollBatch(BucketId bucket,
                         std::optional<std::size_t> limit) const {
  Require(impl_, "PriorityQueue has been moved from");
  const auto [has_limit, native_limit] = NativeLimit(limit);
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_poll_batch(
        *impl_->native, bucket, has_limit, native_limit);
  });
  return OwnedPriorityQueueBatch(
      std::make_unique<OwnedPriorityQueueBatch::Impl>(std::move(native)));
}

BufferResult PriorityQueue::PeekInto(BucketId bucket,
                                     MutableBytesView output) {
  Require(impl_, "PriorityQueue has been moved from");
  return detail::ToBufferResult(detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_peek_into(
        *impl_->native, bucket,
        rust::Slice<Byte>(output.data(), output.size()));
  }));
}

BufferResult PriorityQueue::PollInto(BucketId bucket,
                                     MutableBytesView output) {
  Require(impl_, "PriorityQueue has been moved from");
  return detail::ToBufferResult(detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_poll_into(
        *impl_->native, bucket,
        rust::Slice<Byte>(output.data(), output.size()));
  }));
}

BufferResult
PriorityQueue::PeekBatchInto(BucketId bucket, MutableBytesView output,
                             std::optional<std::size_t> limit) {
  Require(impl_, "PriorityQueue has been moved from");
  const auto [has_limit, native_limit] = NativeLimit(limit);
  return detail::ToBufferResult(detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_peek_batch_into(
        *impl_->native, bucket, has_limit, native_limit,
        rust::Slice<Byte>(output.data(), output.size()));
  }));
}

BufferResult
PriorityQueue::PollBatchInto(BucketId bucket, MutableBytesView output,
                             std::optional<std::size_t> limit) {
  Require(impl_, "PriorityQueue has been moved from");
  const auto [has_limit, native_limit] = NativeLimit(limit);
  return detail::ToBufferResult(detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_poll_batch_into(
        *impl_->native, bucket, has_limit, native_limit,
        rust::Slice<Byte>(output.data(), output.size()));
  }));
}

void PriorityQueue::Advance(BucketId bucket, BytesView key) const {
  Require(impl_, "PriorityQueue has been moved from");
  detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_advance(
        *impl_->native, bucket, detail::RustBytes(key));
  });
}

std::optional<std::vector<Byte>>
PriorityQueue::Cursor(BucketId bucket) const {
  Require(impl_, "PriorityQueue has been moved from");
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_priority_queue_cursor(
        *impl_->native, bucket);
  });
  if (!structured_ffi::native_priority_queue_cursor_has_value(*native))
    return std::nullopt;
  const auto value = detail::Translate([&] {
    return structured_ffi::native_priority_queue_cursor_value(*native);
  });
  if (value.empty())
    return std::vector<Byte>{};
  return std::vector<Byte>(value.begin(), value.end());
}

PriorityQueue Db::NewPriorityQueue(std::string_view name) {
  Require(impl_, "structured Db has been moved from");
  RequireExclusiveOwner(impl_);
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_db_new_priority_queue(
        *impl_->native, detail::RustStr(name));
  });
  PriorityQueue::Impl::Owner owner = impl_;
  return PriorityQueue(std::make_unique<PriorityQueue::Impl>(
      std::move(owner), std::move(native)));
}

PriorityQueue Db::GetPriorityQueue(std::string_view name) const {
  Require(impl_, "structured Db has been moved from");
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_db_get_priority_queue(
        *impl_->native, detail::RustStr(name));
  });
  PriorityQueue::Impl::Owner owner = impl_;
  return PriorityQueue(std::make_unique<PriorityQueue::Impl>(
      std::move(owner), std::move(native)));
}

PriorityQueue Db::GetOrNewPriorityQueue(std::string_view name) {
  Require(impl_, "structured Db has been moved from");
  RequireExclusiveOwner(impl_);
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_db_get_or_new_priority_queue(
        *impl_->native, detail::RustStr(name));
  });
  PriorityQueue::Impl::Owner owner = impl_;
  return PriorityQueue(std::make_unique<PriorityQueue::Impl>(
      std::move(owner), std::move(native)));
}

PriorityQueue SingleDb::NewPriorityQueue(std::string_view name) {
  Require(impl_, "structured SingleDb has been moved from");
  RequireExclusiveOwner(impl_);
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_single_db_new_priority_queue(
        *impl_->native, detail::RustStr(name));
  });
  PriorityQueue::Impl::Owner owner = impl_;
  return PriorityQueue(std::make_unique<PriorityQueue::Impl>(
      std::move(owner), std::move(native)));
}

PriorityQueue SingleDb::GetPriorityQueue(std::string_view name) const {
  Require(impl_, "structured SingleDb has been moved from");
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_single_db_get_priority_queue(
        *impl_->native, detail::RustStr(name));
  });
  PriorityQueue::Impl::Owner owner = impl_;
  return PriorityQueue(std::make_unique<PriorityQueue::Impl>(
      std::move(owner), std::move(native)));
}

PriorityQueue SingleDb::GetOrNewPriorityQueue(std::string_view name) {
  Require(impl_, "structured SingleDb has been moved from");
  RequireExclusiveOwner(impl_);
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_single_db_get_or_new_priority_queue(
        *impl_->native, detail::RustStr(name));
  });
  PriorityQueue::Impl::Owner owner = impl_;
  return PriorityQueue(std::make_unique<PriorityQueue::Impl>(
      std::move(owner), std::move(native)));
}

} // namespace cobble::structured

File: cobble-binding/cobble-cpp/src/structured/priority_queue.rs
use std::sync::Arc;

use bytes::Bytes;
use cobble_binding::structured::{StructuredDb, StructuredSingleDb, ffi as data_structure_ffi};

use crate::structured_bridge::ffi;

use super::conversion::{format_error, input_error};
use super::encoding::{
    CsrbColumns, CsrbRow, STATUS_BUFFER_TOO_SMALL, STATUS_END, STATUS_NOT_FOUND, STATUS_OK, prepare,
};
use super::multi_get::buffer_result;
use super::{BridgeResult, NativeStructuredDb, NativeStructuredSingleDb};

fn invalid_state(message: impl std::fmt::Display) -> String {
    format!("CB_INVALID_STATE: {message}")
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum PriorityQueueOperation {
    Peek,
    Poll,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
struct PriorityQueueRequest {
    operation: PriorityQueueOperation,
    bucket: u16,
    limit: Option<usize>,
    single: bool,
}

struct PendingPriorityQueueBatch {
    request: PriorityQueueRequest,
    rows: Vec<(Bytes, Bytes)>,
}

enum NativePriorityQueueOwner {
    Db(Arc<StructuredDb>),
    SingleDb(Arc<StructuredSingleDb>),
}

pub(crate) struct NativeStructuredPriorityQueue {
    descriptor: data_structure_ffi::PriorityQueueDescriptor,
    pending: Option<PendingPriorityQueueBatch>,
    // Declared last so all descriptor/pending data is released before the last
    // database owner. This prevents close-on-drop from waiting on a child.
    owner: NativePriorityQueueOwner,
}

pub(crate) struct NativePriorityQueueBatch {
    rows: Vec<(Bytes, Bytes)>,
}

pub(crate) struct NativePriorityQueueCursor {
    value: Option<Vec<u8>>,
}

impl NativeStructuredPriorityQueue {
    fn ensure_idle(&self) -> BridgeResult<()> {
        if self.pending.is_some() {
            Err(invalid_state(
                "priority queue has a pending caller-buffer result; retry the same operation",
            ))
        } else {
            Ok(())
        }
    }

    fn offer(&self, bucket: u16, key: &[u8], value: &[u8]) -> BridgeResult<()> {
        self.ensure_idle()?;
        match &self.owner {
            NativePriorityQueueOwner::Db(db) => data_structure_ffi::db_priority_queue_offer(
                db,
                &self.descriptor,
                bucket,
                key,
                value,
            ),
            NativePriorityQueueOwner::SingleDb(db) => {
                data_structure_ffi::single_db_priority_queue_offer(
                    db,
                    &self.descriptor,
                    bucket,
                    key,
                    value,
                )
            }
        }
        .map_err(format_error)
    }

    fn delete(&self, bucket: u16, key: &[u8]) -> BridgeResult<()> {
        self.ensure_idle()?;
        match &self.owner {
            NativePriorityQueueOwner::Db(db) => {
                data_structure_ffi::db_priority_queue_delete(db, &self.descriptor, bucket, key)
            }
            NativePriorityQueueOwner::SingleDb(db) => {
                data_structure_ffi::single_db_priority_queue_delete(
                    db,
                    &self.descriptor,
                    bucket,
                    key,
                )
            }
        }
        .map_err(format_error)
    }

    fn peek_batch(&self, bucket: u16, limit: Option<usize>) -> BridgeResult<Vec<(Bytes, Bytes)>> {
        match &self.owner {
            NativePriorityQueueOwner::Db(db) => data_structure_ffi::db_priority_queue_peek_batch(
                db,
                &self.descriptor,
                bucket,
                limit,
            ),
            NativePriorityQueueOwner::SingleDb(db) => {
                data_structure_ffi::single_db_priority_queue_peek_batch(
                    db,
                    &self.descriptor,
                    bucket,
                    limit,
                )
            }
        }
        .map_err(format_error)
    }

    fn advance_internal(&self, bucket: u16, key: &[u8]) -> BridgeResult<()> {
        match &self.owner {
            NativePriorityQueueOwner::Db(db) => {
                data_structure_ffi::db_priority_queue_advance(db, &self.descriptor, bucket, key)
            }
            NativePriorityQueueOwner::SingleDb(db) => {
                data_structure_ffi::single_db_priority_queue_advance(
                    db,
                    &self.descriptor,
                    bucket,
                    key,
                )
            }
        }
        .map_err(format_error)
    }

    fn cursor_internal(&self, bucket: u16) -> BridgeResult<Option<Vec<u8>>> {
        match &self.owner {
            NativePriorityQueueOwner::Db(db) => {
                data_structure_ffi::db_priority_queue_cursor(db, &self.descriptor, bucket)
            }
            NativePriorityQueueOwner::SingleDb(db) => {
                data_structure_ffi::single_db_priority_queue_cursor(db, &self.descriptor, bucket)
            }
        }
        .map_err(format_error)
    }

    fn owned_batch(
        &self,
        request: PriorityQueueRequest,
    ) -> BridgeResult<Box<NativePriorityQueueBatch>> {
        self.ensure_idle()?;
        let rows = self.peek_batch(request.bucket, request.limit)?;
        if request.operation == PriorityQueueOperation::Poll
            && let Some((key, _)) = rows.last()
        {
            self.advance_internal(request.bucket, key)?;
        }
        Ok(Box::new(NativePriorityQueueBatch { rows }))
    }

    fn batch_into(
        &mut self,
        request: PriorityQueueRequest,
        output: &mut [u8],
    ) -> BridgeResult<ffi::NativeBufferResult> {
        match self.pending.as_ref() {
            Some(pending) if pending.request != request => {
                return Err(invalid_state(
                    "priority queue caller-buffer retry must use the same operation, bucket, and limit",
                ));
            }
            Some(_) => {}
            None => {
                let rows = self.peek_batch(request.bucket, request.limit)?;
                self.pending = Some(PendingPriorityQueueBatch { request, rows });
            }
        }

        let pending = self.pending.as_ref().expect("pending batch initialized");
        let csrb_rows = pending
            .rows
            .iter()
            .map(|(key, value)| CsrbRow {
                bucket: request.bucket,
                key,
                columns: CsrbColumns::PriorityQueue(value),
            })
            .collect::<Vec<_>>();
        let prepared = prepare(&csrb_rows, pending.rows.is_empty(), false)?;
        let required = prepared.required_len();
        if output.len() < required {
            return Ok(buffer_result(
                STATUS_BUFFER_TOO_SMALL,
                0,
                required,
                pending.rows.len(),
            ));
        }

        // The buffer is still untouched here. Complete the externally visible
        // poll first so an advance failure cannot expose rows as consumed.
        if request.operation == PriorityQueueOperation::Poll
            && let Some((key, _)) = pending.rows.last()
        {
            self.advance_internal(request.bucket, key)?;
        }
        let written = prepared.encode_into(output);
        let row_count = pending.rows.len();
        self.pending = None;
        Ok(buffer_result(
            if row_count == 0 {
                if request.single {
                    STATUS_NOT_FOUND
                } else {
                    STATUS_END
                }
            } else {
                STATUS_OK
            },
            written,
            written,
            row_count,
        ))
    }
}

fn exclusive_db(db: &mut NativeStructuredDb) -> BridgeResult<&mut StructuredDb> {
    Arc::get_mut(&mut db.db).ok_or_else(|| {
        invalid_state(
            "priority queue creation requires exclusive ownership; release every structured scan cursor, schema builder, and priority queue first",
        )
    })
}

fn exclusive_single_db(db: &mut NativeStructuredSingleDb) -> BridgeResult<&mut StructuredSingleDb> {
    Arc::get_mut(&mut db.db).ok_or_else(|| {
        invalid_state(
            "priority queue creation requires exclusive ownership; release every structured scan cursor, schema builder, and priority queue first",
        )
    })
}

fn native_queue(
    descriptor: data_structure_ffi::PriorityQueueDescriptor,
    owner: NativePriorityQueueOwner,
) -> Box<NativeStructuredPriorityQueue> {
    Box::new(NativeStructuredPriorityQueue {
        descriptor,
        pending: None,
        owner,
    })
}

pub(crate) fn native_structured_db_new_priority_queue(
    db: &mut NativeStructuredDb,
    name: &str,
) -> BridgeResult<Box<NativeStructuredPriorityQueue>> {
    let descriptor =
        data_structure_ffi::db_new_priority_queue_descriptor(exclusive_db(db)?, name.to_owned())
            .map_err(format_error)?;
    Ok(native_queue(
        descriptor,
        NativePriorityQueueOwner::Db(Arc::clone(&db.db)),
    ))
}

pub(crate) fn native_structured_db_get_priority_queue(
    db: &NativeStructuredDb,
    name: &str,
) -> BridgeResult<Box<NativeStructuredPriorityQueue>> {
    let descriptor = data_structure_ffi::db_get_priority_queue_descriptor(&db.db, name.to_owned())
        .map_err(format_error)?;
    Ok(native_queue(
        descriptor,
        NativePriorityQueueOwner::Db(Arc::clone(&db.db)),
    ))
}

pub(crate) fn native_structured_db_get_or_new_priority_queue(
    db: &mut NativeStructuredDb,
    name: &str,
) -> BridgeResult<Box<NativeStructuredPriorityQueue>> {
    let descriptor = data_structure_ffi::db_get_or_new_priority_queue_descriptor(
        exclusive_db(db)?,
        name.to_owned(),
    )
    .map_err(format_error)?;
    Ok(native_queue(
        descriptor,
        NativePriorityQueueOwner::Db(Arc::clone(&db.db)),
    ))
}

pub(crate) fn native_structured_single_db_new_priority_queue(
    db: &mut NativeStructuredSingleDb,
    name: &str,
) -> BridgeResult<Box<NativeStructuredPriorityQueue>> {
    let descriptor = data_structure_ffi::single_db_new_priority_queue_descriptor(
        exclusive_single_db(db)?,
        name.to_owned(),
    )
    .map_err(format_error)?;
    Ok(native_queue(
        descriptor,
        NativePriorityQueueOwner::SingleDb(Arc::clone(&db.db)),
    ))
}

pub(crate) fn native_structured_single_db_get_priority_queue(
    db: &NativeStructuredSingleDb,
    name: &str,
) -> BridgeResult<Box<NativeStructuredPriorityQueue>> {
    let descriptor =
        data_structure_ffi::single_db_get_priority_queue_descriptor(&db.db, name.to_owned())
            .map_err(format_error)?;
    Ok(native_queue(
        descriptor,
        NativePriorityQueueOwner::SingleDb(Arc::clone(&db.db)),
    ))
}

pub(crate) fn native_structured_single_db_get_or_new_priority_queue(
    db: &mut NativeStructuredSingleDb,
    name: &str,
) -> BridgeResult<Box<NativeStructuredPriorityQueue>> {
    let descriptor = data_structure_ffi::single_db_get_or_new_priority_queue_descriptor(
        exclusive_single_db(db)?,
        name.to_owned(),
    )
    .map_err(format_error)?;
    Ok(native_queue(
        descriptor,
        NativePriorityQueueOwner::SingleDb(Arc::clone(&db.db)),
    ))
}

pub(crate) fn native_structured_priority_queue_column_family(
    queue: &NativeStructuredPriorityQueue,
) -> &str {
    queue.descriptor.column_family()
}

pub(crate) fn native_structured_priority_queue_offer(
    queue: &NativeStructuredPriorityQueue,
    bucket: u16,
    key: &[u8],
    value: &[u8],
) -> BridgeResult<()> {
    queue.offer(bucket, key, value)
}

pub(crate) fn native_structured_priority_queue_delete(
    queue: &NativeStructuredPriorityQueue,
    bucket: u16,
    key: &[u8],
) -> BridgeResult<()> {
    queue.delete(bucket, key)
}

fn request(
    operation: PriorityQueueOperation,
    bucket: u16,
    has_limit: bool,
    limit: usize,
    single: bool,
) -> PriorityQueueRequest {
    PriorityQueueRequest {
        operation,
        bucket,
        limit: has_limit.then_some(limit),
        single,
    }
}

pub(crate) fn native_structured_priority_queue_peek(
    queue: &NativeStructuredPriorityQueue,
    bucket: u16,
) -> BridgeResult<Box<NativePriorityQueueBatch>> {
    queue.owned_batch(request(PriorityQueueOperation::Peek, bucket, true, 1, true))
}

pub(crate) fn native_structured_priority_queue_poll(
    queue: &NativeStructuredPriorityQueue,
    bucket: u16,
) -> BridgeResult<Box<NativePriorityQueueBatch>> {
    queue.owned_batch(request(PriorityQueueOperation::Poll, bucket, true, 1, true))
}

pub(crate) fn native_structured_priority_queue_peek_batch(
    queue: &NativeStructuredPriorityQueue,
    bucket: u16,
    has_limit: bool,
    limit: usize,
) -> BridgeResult<Box<NativePriorityQueueBatch>> {
    queue.owned_batch(request(
        PriorityQueueOperation::Peek,
        bucket,
        has_limit,
        limit,
        false,
    ))
}

pub(crate) fn native_structured_priority_queue_poll_batch(
    queue: &NativeStructuredPriorityQueue,
    bucket: u16,
    has_limit: bool,
    limit: usize,
) -> BridgeResult<Box<NativePriorityQueueBatch>> {
    queue.owned_batch(request(
        PriorityQueueOperation::Poll,
        bucket,
        has_limit,
        limit,
        false,
    ))
}

pub(crate) fn native_structured_priority_queue_peek_into(
    queue: &mut NativeStructuredPriorityQueue,
    bucket: u16,
    output: &mut [u8],
) -> BridgeResult<ffi::NativeBufferResult> {
    queue.batch_into(
        request(PriorityQueueOperation::Peek, bucket, true, 1, true),
        output,
    )
}

pub(crate) fn native_structured_priority_queue_poll_into(
    queue: &mut NativeStructuredPriorityQueue,
    bucket: u16,
    output: &mut [u8],
) -> BridgeResult<ffi::NativeBufferResult> {
    queue.batch_into(
        request(PriorityQueueOperation::Poll, bucket, true, 1, true),
        output,
    )
}

pub(crate) fn native_structured_priority_queue_peek_batch_into(
    queue: &mut NativeStructuredPriorityQueue,
    bucket: u16,
    has_limit: bool,
    limit: usize,
    output: &mut [u8],
) -> BridgeResult<ffi::NativeBufferResult> {
    queue.batch_into(
        request(
            PriorityQueueOperation::Peek,
            bucket,
            has_limit,
            limit,
            false,
        ),
        output,
    )
}

pub(crate) fn native_structured_priority_queue_poll_batch_into(
    queue: &mut NativeStructuredPriorityQueue,
    bucket: u16,
    has_limit: bool,
    limit: usize,
    output: &mut [u8],
) -> BridgeResult<ffi::NativeBufferResult> {
    queue.batch_into(
        request(
            PriorityQueueOperation::Poll,
            bucket,
            has_limit,
            limit,
            false,
        ),
        output,
    )
}

pub(crate) fn native_structured_priority_queue_advance(
    queue: &NativeStructuredPriorityQueue,
    bucket: u16,
    key: &[u8],
) -> BridgeResult<()> {
    queue.ensure_idle()?;
    queue.advance_internal(bucket, key)
}

pub(crate) fn native_structured_priority_queue_cursor(
    queue: &NativeStructuredPriorityQueue,
    bucket: u16,
) -> BridgeResult<Box<NativePriorityQueueCursor>> {
    queue.ensure_idle()?;
    Ok(Box::new(NativePriorityQueueCursor {
        value: queue.cursor_internal(bucket)?,
    }))
}

pub(crate) fn native_priority_queue_batch_size(batch: &NativePriorityQueueBatch) -> usize {
    batch.rows.len()
}

pub(crate) fn native_priority_queue_batch_key(
    batch: &NativePriorityQueueBatch,
    index: usize,
) -> BridgeResult<&[u8]> {
    batch
        .rows
        .get(index)
        .map(|(key, _)| key.as_ref())
        .ok_or_else(|| input_error("priority queue batch index is out of bounds"))
}

pub(crate) fn native_priority_queue_batch_value(
    batch: &NativePriorityQueueBatch,
    index: usize,
) -> BridgeResult<&[u8]> {
    batch
        .rows
        .get(index)
        .map(|(_, value)| value.as_ref())
        .ok_or_else(|| input_error("priority queue batch index is out of bounds"))
}

pub(crate) fn native_priority_queue_cursor_has_value(cursor: &NativePriorityQueueCursor) -> bool {
    cursor.value.is_some()
}

pub(crate) fn native_priority_queue_cursor_value(
    cursor: &NativePriorityQueueCursor,
) -> BridgeResult<&[u8]> {
    cursor
        .value
        .as_deref()
        .ok_or_else(|| input_error("priority queue cursor is empty"))
}

#[cfg(test)]
mod tests {
    use cobble_binding::{Config, VolumeDescriptor};

    use super::*;

    #[test]
    fn caller_buffer_poll_retry_is_transactional() {
        let root = std::env::temp_dir().join(format!(
            "cobble_cpp_priority_queue_retry_{}_{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_nanos()
        ));
        let config = Config {
            volumes: VolumeDescriptor::single_volume(format!("file://{}", root.to_string_lossy())),
            total_buckets: 1,
            ..Config::default()
        };
        let db = StructuredSingleDb::open(config).unwrap();
        let mut native_db = NativeStructuredSingleDb { db: Arc::new(db) };
        let mut queue =
            native_structured_single_db_new_priority_queue(&mut native_db, "jobs").unwrap();
        native_structured_priority_queue_offer(&queue, 0, b"k1", b"v1").unwrap();
        native_structured_priority_queue_offer(&queue, 0, b"k2", b"v2").unwrap();

        let mut small = [0xa5; 4];
        let before = small;
        let first =
            native_structured_priority_queue_poll_batch_into(&mut queue, 0, true, 1, &mut small)
                .unwrap();
        assert_eq!(first.status, STATUS_BUFFER_TOO_SMALL);
        assert_eq!(small, before);
        assert!(native_structured_priority_queue_peek(&queue, 0).is_err());
        assert!(
            native_structured_priority_queue_poll_batch_into(
                &mut queue,
                0,
                true,
                2,
                &mut vec![0; first.bytes_required as usize],
            )
            .is_err()
        );

        let mut output = vec![0; first.bytes_required as usize];
        let retry =
            native_structured_priority_queue_poll_batch_into(&mut queue, 0, true, 1, &mut output)
                .unwrap();
        assert_eq!(retry.status, STATUS_OK);
        assert_eq!(retry.row_count, 1);
        assert_eq!(&output[..4], b"CSRB");
        assert_eq!(
            native_structured_priority_queue_cursor(&queue, 0)
                .unwrap()
                .value
                .as_deref(),
            Some(b"k1".as_slice())
        );
        let next = native_structured_priority_queue_peek(&queue, 0).unwrap();
        assert_eq!(next.rows[0].0.as_ref(), b"k2");

        drop(queue);
        native_db.db.close().unwrap();
        drop(native_db);
        let _ = std::fs::remove_dir_all(root);
    }
}
